tests/syntax/test_specifiers.py

```diff
@@ -47,181 +52,501 @@ def test_lazy_value_in_requirement():
+
+def test_offset_along_3d():
+    ego = sampleEgoFrom(
+        'ego = new Object at (10, 40, 20)\n'
+        'ego = new Object offset along (90 deg, 0, 90 deg) by (-10, 5, 20)'
+    )
+    assert tuple(ego.position) == pytest.approx((5, 60, 30))
+
+    ego = sampleEgoFrom(
+        'ego = new Object at (10, 40, 20)\n'
+        'ego = new Object offset along (90 deg, 90 deg, 90 deg) by (-10, 5, 20)'
     )
-    assert tuple(ego.position) == pytest.approx((15, 50))
+    assert tuple(ego.position) == pytest.approx((20, 60, 25))
 
 def test_offset_along_no_ego():
-    with pytest.raises(RuntimeParseError):
-        compileScenic('ego = Object offset along 0 by 10 @ 0')
+    with pytest.raises(InvalidScenarioError):
+        compileScenic('ego = new Object offset along 0 by 10 @ 0')
 
+# Left/Right Of 
 def test_left_of_vector():
-    ego = sampleEgoFrom('ego = Object left of 10 @ 20, facing 90 deg')
-    assert tuple(ego.position) == pytest.approx((10, 19.5))
-    ego = sampleEgoFrom('ego = Object left of 10 @ 20, with width 10')
-    assert tuple(ego.position) == pytest.approx((5, 20))
+    ego = sampleEgoFrom('ego = new Object left of 10 @ 20, facing 90 deg')
+    assert tuple(ego.position) == pytest.approx((10, 19.5, 0))
+    ego = sampleEgoFrom('ego = new Object left of 10 @ 20, with width 10')
+    assert tuple(ego.position) == pytest.approx((5, 20, 0))
+
+def test_left_of_vector_3d():
+    ego = sampleEgoFrom('ego = new Object left of (10, 20, 15), facing (90 deg, 0, 90 deg)')
+    assert tuple(ego.position) == pytest.approx((10, 20, 15.5))
 
 def test_left_of_vector_by():
-    ego = sampleEgoFrom('ego = Object left of 10 @ 20 by 20')
-    assert tuple(ego.position) == pytest.approx((-10.5, 20))
-    ego = sampleEgoFrom('ego = Object left of 10 @ 20 by 20 @ 5')
-    assert tuple(ego.position) == pytest.approx((-10.5, 25))
+    ego = sampleEgoFrom('ego = new Object left of 10 @ 20 by 20')
+    assert tuple(ego.position) == pytest.approx((-10.5, 20, 0))
+    ego = sampleEgoFrom('ego = new Object left of 10 @ 20 by 20 @ 5')
+    assert tuple(ego.position) == pytest.approx((-10.5, 25, 0))
+
+def test_left_of_vector_by_3d():
+    ego = sampleEgoFrom('ego = new Object left of (10, 20, 15) by 20')
+    assert tuple(ego.position) == pytest.approx((-10.5, 20, 15))
+
+def test_left_of_oriented_point():
+    ego = sampleEgoFrom(
+                        'target = new OrientedPoint facing (-90 deg, 0, 0)\n'
+                        'ego = new Object left of target'
+                       )
+    assert tuple(ego.position) == pytest.approx((0, 0.5, 0), abs=1e-6)
+    assert ego.orientation.approxEq(Orientation.fromEuler(-math.pi/2, 0, 0))
+
+def test_left_of_object():
+    ego = sampleEgoFrom(
+                        'target = new Object facing (-90 deg, 0, 0)\n'
+                        'ego = new Object left of target'
+                       )
+    assert tuple(ego.position) == pytest.approx((0, 1, 0), abs=2*ego.contactTolerance)
```

Comment:
I wouldn't use the `pytest.approx` tolerance in this way: if the Y coordinate were *less* than 1 by the tolerance, that would be a bug! I think it's better to assert approximate equality with the expected value, which here is `(0, 1+ego.contactTolerance, 0)` (I think -- I don't see why we would should the objects apart by twice the contact tolerance).